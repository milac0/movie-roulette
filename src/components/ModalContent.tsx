import React, { useEffect, useState, Fragment } from "react";
import axios from "axios";
import { css } from "emotion";
import { colors } from "../theme/theme";
import { Link } from "react-router-dom";
import { randomNumber } from "../helpers";

const modalRandom = css`
  & .poster {
    height: 370.5px;
    margin-bottom: 1em;
    width: 250px;
    & img {
      height: auto;
      width: 100%;
    }
  }
  & div {
    text-align: center;
    & .nextBtn {
      background: ${colors.secondary};
      border: none;
      border-radius: 5px;
      cursor: pointer;
      font-weight: 700;
      padding: 0.5em 1em;
      text-align: center;
      &:active {
        transform: translateY(2px);
      }
    }
  }
`;

const randomSkeleton = css`
  height: 370.5px;
  width: 250px;
`;

interface Props {}

const ModalContent: React.FC<Props> = () => {
  const [movie, setMovie] = useState(null);
  const [next, setNext] = useState(0);
  useEffect(() => {
    (async () => {
      const totalPages = (
        await axios.get(
          `/discover/movie?api_key=${process.env.API_KEY}&language=en-US&include_adult=false&include_video=false&page=1`
        )
      ).data.total_pages;
      const randomPage = randomNumber(0, totalPages);
      const randomMovie = randomNumber(0, 19);
      const movie = (
        await axios.get(
          `/discover/movie?api_key=${process.env.API_KEY}&language=en-US&include_adult=false&include_video=false&page=${randomPage}`
        )
      ).data.results[randomMovie];
      setMovie(movie);
    })();
  }, [next]);
  return (
    <Fragment>
      {movie === null ? (
        <div className={randomSkeleton}></div>
      ) : (
        <div className={modalRandom}>
          <Link to={`/movies/${movie.id}`}>
            <div className={"poster"}>
              <img
                src={`https://image.tmdb.org/t/p/w500${movie.poster_path}`}
              />
            </div>
          </Link>
          <div>
            <button className="nextBtn" onClick={() => setNext(next + 1)}>
              Random
            </button>
          </div>
        </div>
      )}
    </Fragment>
  );
};

export default ModalContent;
