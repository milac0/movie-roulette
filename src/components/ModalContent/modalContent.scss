@import "../../theme/variables";

.content {
  background: $backgroundGrey;
  border: "2px solid #000";
  height: 430px;
  left: 50%;
  margin: 0;
  padding: 1em;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);

  .close {
    border-radius: 50%;
    cursor: pointer;
    position: absolute;
    right: 0;
    top: 0;
  }

  .poster {
    align-items: center;
    display: flex;
    height: 370.5px;
    margin-top: 0.5em;
    margin-bottom: 1em;
    width: 250px;

    img {
      height: auto;
      width: 100%;
    }
  }

  .options {
    display: flex;
    justify-content: space-around;

    select {
      border: 1px solid $primary;
      border-radius: 5px;
      font-weight: 700;
      padding: 0.5em 0.5em;
      width: 10em;

      option {
        font-weight: 700;
      }
    }

    button {
      background: $secondary;
      border: none;
      border-radius: 5px;
      cursor: pointer;
      font-weight: 700;
      padding: 0.5em 2em;
      text-align: center;
      &:active {
        transform: translateY(2px);
      }
    }
  }
}
