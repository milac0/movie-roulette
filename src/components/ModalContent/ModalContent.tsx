import React, { useEffect, useState } from "react";
import axios from "axios";
import { Link } from "react-router-dom";
import { randomNumber } from "./modalContent.helpers";
import Genre from "../../interface/Genre";
import questionImg from "../../assets/images/question.png";
import closeImg from "../../assets/images/close.png";
import styles from "./modalContent.scss";

interface Props {
  handleClose: () => void;
}

const ModalContent: React.FC<Props> = ({ handleClose }) => {
  const genreURL = `/genre/movie/list?api_key=${process.env.API_KEY}&language=en-US`;
  const getGenreMovieURL = (page: number, genre: string) =>
    `/discover/movie?api_key=${process.env.API_KEY}&language=en-US&include_adult=false&include_video=false&page=${page}&with_genres=${genre}`;

  const [movie, setMovie] = useState(null);
  const [genres, setGenres] = useState<Array<Genre>>([]);
  const [select, setSelect] = useState("28");
  const [firstRender, setFirstRender] = useState(true);
  const [next, setNext] = useState(0);

  useEffect(() => {
    (async () => {
      try {
        const {
          data: { genres: genresList }
        } = await axios.get(genreURL);
        setGenres(genresList);
      } catch (err) {
        console.log(err);
      }
    })();
  }, []);

  useEffect(() => {
    (async () => {
      if (!firstRender) {
        // total pages is always 500
        const totalPages = 500;
        const randomPage = randomNumber(1, totalPages);
        const randomMovie = randomNumber(0, 19);
        try {
          const {
            data: { results: movies }
          } = await axios.get(getGenreMovieURL(randomPage, select));
          setMovie(movies[randomMovie]);
        } catch (err) {
          console.log(err);
        }
      }
    })();
  }, [next]);

  const handleChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    setSelect(e.target.value);
  };

  const handleNext = () => {
    setFirstRender(false);
    setNext(next + 1);
  };

  return (
    <div className={styles.content}>
      <img className={styles.close} src={closeImg} onClick={handleClose} />
      <div className={styles.poster}>
        {firstRender ? (
          <img src={questionImg} />
        ) : movie === null ? null : (
          <Link to={`/movies/${movie.id}`}>
            {movie.poster_path !== null ? (
              <img
                src={`https://image.tmdb.org/t/p/w500${movie.poster_path}`}
              />
            ) : (
              <img src={questionImg} />
            )}
          </Link>
        )}
      </div>
      <div className={styles.options}>
        {genres.length > 0 ? (
          <select onChange={handleChange} value={select}>
            {genres.map((genre: Genre) => {
              return (
                <option key={genre.id} value={genre.id}>
                  {genre.name}
                </option>
              );
            })}
          </select>
        ) : (
          <select>Loading</select>
        )}
        <button onClick={handleNext}>Random</button>
      </div>
    </div>
  );
};

export default ModalContent;
