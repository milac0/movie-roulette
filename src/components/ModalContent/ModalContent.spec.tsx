import React from 'react';
import renderer from 'react-test-renderer';

import ModalContent from './ModalContent';

describe('ModalContent renders correctly', () => {
    test('snapshot renders', () => {
        const handleClose = jest.fn();

        const tree = renderer.create(<ModalContent handleClose={handleClose} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

