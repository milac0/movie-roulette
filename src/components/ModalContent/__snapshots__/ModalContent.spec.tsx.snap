// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalContent renders correctly snapshot renders 1`] = `
<div
  className="content"
>
  <img
    className="close"
    onClick={[MockFunction]}
    src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+P+/HgAFhAJ/wlseKgAAAABJRU5ErkJggg=="
  />
  <div
    className="poster"
  >
    <img
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+P+/HgAFhAJ/wlseKgAAAABJRU5ErkJggg=="
    />
  </div>
  <div
    className="options"
  >
    <select>
      Loading
    </select>
    <button
      onClick={[Function]}
    >
      Random
    </button>
  </div>
</div>
`;
